Support fused batchnorm with any ndims and axis

benbarsdell opened the pull request:
- Modifies the BatchNormalizationBase layer to support fused=True with any number of dimensions and any (single) axis by adding reshape operations around the calls to nn.fused_batch_norm (which itself only supports NCHW and NHWC tensors).
- Fixes tests to account for the broader support and adds a new test for a 3D convnet.
- Use of fused=True is particularly important for mixed-precision training.

cc @reedwm @nluehr

benbarsdell committed:
Support fused batchnorm with any ndims and axis

- Modifies the BatchNormalizationBase layer to support fused=True with
  and number of dimensions and any (single) axis by adding reshape
  operations around the calls to nn.fused_batch_norm (which itself only
  supports NCHW and NHWC tensors).
- Fixes tests to account for the broader support and adds a new test for
  a 3D convnet.

reedwm commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -281,6 +273,51 @@ def _support_zero_size_input(self):
+      # Merge excess spatial dims into the second spatial dim.
+      # Transform [N, C, H, W, ...] to [N, C, H, product(W, ...)].
+      # Or        [N, H, W, ..., C] to [N, H, product(W, ...), C].
+      merge_dim = 2 if is_channels_last else 3
```
Similar to above, maybe do:

```python
product = 1
for elem in shape[merge_dim:merg_dim + ndims - 4]:
  product *= elem
shape[merge_dim:merg_dim + ndims - 4] = [product]
```

benbarsdell replied:
Done.

reedwm commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -315,18 +352,22 @@ def build(self, input_shape):
+        # reshape the input/output tensor to/from an equivalent 4D shape.
+        fused_shape, fused_axis = self._get_shape_and_axis_for_fused(
+            input_shape.dims, self.axis[0])
+        fused_shape = tensor_shape.TensorShape(fused_shape)
```
why convert this to a TensorShape then immediately convert back to a list?

benbarsdell replied:
Done. (Twas a remnant of a fight I had initially with TensorShape vs. list, None vs. -1 etc.).

reedwm commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -315,18 +352,22 @@ def build(self, input_shape):
+        fused_shape, fused_axis = self._get_shape_and_axis_for_fused(
+            input_shape.dims, self.axis[0])
+        fused_shape = tensor_shape.TensorShape(fused_shape)
+        self._input_fused_shape = [-1] + fused_shape.as_list()[1:]
```
The issue with computing this in `build()` is that the input shape might change between calls to the layer. The channel dimension must stay the same since that affects the size of the weights, but other dimensions can differ.

Instead, compute this in `call`.

benbarsdell replied:
Done.

reedwm commented on tensorflow/python/keras/layers/normalization_test.py:
```diff
@@ -138,6 +138,27 @@ def test_batchnorm_convnet_channel_last(self):
     np.testing.assert_allclose(np.std(out, axis=(0, 1, 2)), 1.0, atol=1e-1)
 
+  @keras_parameterized.run_all_keras_modes
+  def test_batchnorm_convnet_channel_last_3d_fused(self):
```
I would also add a test with simple hand-written inputs. No need to test gradients. I'm worried this won't catch errors in `_get_shape_and_axis_for_fused`. Note gamma and beta default to ones and zeros, so you can effectively ignore them in tests that don't take gradients.

benbarsdell replied:
Done.

reedwm reviewed (changes requested):
Thanks for the PR!

benbarsdell committed:
Apply initial corrections from PR 40338 review

- Simplify condition.
- Improve and format docstring.
- Use simpler shape manipulation logic.
- Remove redundant shape -> list transform.

benbarsdell committed:
Apply remaining corrections from PR 40338 review

- Moves fused shape calculation from build() to call() so that it
  handles dynamic input shapes.
- Adds a new test that checks for specific results and ensures that
  dynamic input shapes work.

benbarsdell commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -315,18 +352,22 @@ def build(self, input_shape):
 
+    fused_axis = self.axis
+    self._input_fused_shape = None
     if self.fused in (None, True):
```
Done.

benbarsdell commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -315,18 +352,22 @@ def build(self, input_shape):
+        self._input_fused_shape = [-1] + fused_shape.as_list()[1:]
+        fused_axis = [fused_axis]
+
+      if not self._USE_V2_BEHAVIOR:
```
Done.

benbarsdell commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -281,6 +273,51 @@ def _support_zero_size_input(self):
         'experimental_enable_get_next_as_optional', False)
 
+  def _get_shape_and_axis_for_fused(self, nd_shape, nd_axis):
+    '''Returns a 4D shape and axis (1 or 3) to which nd_shape and nd_axis can
```
Done.

benbarsdell commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -281,6 +273,51 @@ def _support_zero_size_input(self):
+    '''Returns a 4D shape and axis (1 or 3) to which nd_shape and nd_axis can
+    be changed without changing the result of the batch normalization operation.
+    '''
+    assert(isinstance(nd_axis, int))
```
`isinstance(axis, int)` is already used in several places in this class, including in an error check at the beginning of the `__init__` function.

benbarsdell commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -281,6 +273,51 @@ def _support_zero_size_input(self):
+      else:
+        # Merge excess pre-axis dims into first dim.
+        # Transform [N, ..., C, ...] to [product(N, ...), C, ...].
+        for dim in range(axis - 1, 0, -1):
```
Done.

gbaned commented:
@benbarsdell Can you please check @reedwm's comments and keep us posted. Thanks!

benbarsdell committed:
Apply more corrections from PR 40338 review

- Remove unneeded variable.
- Do 4D shape computation in-graph so that unknown shapes are supported
  in graph mode.
- Compute test target values on the fly so that it's clear where they
  come from.
- Add a test to check that unknown shapes are supported in graph mode.

benbarsdell commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -315,33 +365,9 @@ def build(self, input_shape):
-      else:
-        raise ValueError('Unsupported axis, fused batch norm only supports '
-                         'axis == [1] or axis == [3]')
+    self._input_fused_shape = None
```
Fixed.

benbarsdell commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -499,6 +525,27 @@ def _fused_batch_norm(self, inputs, training):
 
+    original_shape = None
+    fused_axis = self.axis[0]
+    input_shape = inputs.shape.as_list()
```
Done.

benbarsdell commented on tensorflow/python/keras/layers/normalization_test.py:
```diff
@@ -153,11 +153,48 @@ def test_batchnorm_convnet_channel_last_3d_fused(self):
+    moving_mean = keras.backend.eval(norm.moving_mean)
+    moving_variance = keras.backend.eval(norm.moving_variance)
+    np.testing.assert_allclose(
+        moving_mean, np.array([936., 937., 938.]), rtol=1e-5)
```
I've changed it to compute the target values on the fly so that it's clear where they come from.

benbarsdell commented:
I managed to root-cause the `layer_correctness_test` failure. There appears to be a bug [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L60) (also [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L229)) where the input is interpreted as `OutputT` instead of `InputT`.
I'm not sure how to add a tensor typecast there, but just removing that piece of code does fix the problem.

Let me know how you want to deal with it.

reedwm commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -281,53 +281,55 @@ def _get_shape_and_axis_for_fused(self, nd_shape, nd_axis):
-        for dim in shape[:axis]:
-          product *= dim
-        shape[:axis] = [product]
+        product = math_ops.reduce_prod(shape[:axis])
```
If you pass `keepdims=True` here, I think you don't need to call `array_ops.reshape` on the line below. This makes it slightly simpler.

And same for the place below you call `math_ops.reduce_prod`

benbarsdell replied:
Done.

reedwm commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -527,15 +528,18 @@ def _fused_batch_norm(self, inputs, training):
-      original_shape = [-1] + input_shape[1:]
-      inputs = array_ops.reshape(inputs, [-1] + fused_shape[1:])
+      original_shape = array_ops.concat(
+          [constant_op.constant([-1]), input_shape[1:]], axis=0)
```
Can you pass `[-1]` here instead of `constant_op.constant([-1])`? The former is slightly simpler.

benbarsdell replied:
Removed.

reedwm commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -527,15 +528,18 @@ def _fused_batch_norm(self, inputs, training):
+      original_shape = array_ops.concat(
+          [constant_op.constant([-1]), input_shape[1:]], axis=0)
+      fused_shape = array_ops.concat(
+          [constant_op.constant([-1]), fused_shape[1:]], axis=0)
```
Why do you need to replace the first element of `fused_shape` with -1?

benbarsdell replied:
Good point, I've removed these operations.

reedwm commented on tensorflow/python/keras/layers/normalization_test.py:
```diff
@@ -213,7 +277,7 @@ def call(self, x, training):
 
       for _ in range(10):
-        x = constant_op.constant(0.5, shape=[1, 1])
+        x = constant_op.constant(0.5, shape=[2, 1])
```
Why this change?

benbarsdell replied:
The reason is... complicated :)

First, Bessel's correction (`N/(N-1)`) is used for the fused implementation but _not_ for the non-fused. There is a discussion of this [here](https://github.com/keras-team/keras/pull/9028).

Then, for the fused case, the [CPU implementation avoids returning NaN when N < 2](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fused_batch_norm_op.cc#L176-L180), while the GPU implementation (CUDNN) explicitly returns NaN in these cases.

So when the test calls batchnorm with N = 1 and it (with this PR) calls the fused implementation, it returns NaN (on the GPU) and the test fails.
Changing the test size to N = 2 avoids this problem.

This probably isn't ever observed in the real world because batchnorm doesn't work when N is small anyway.

Let me know if you'd like me to take another action here.

rmlarsen replied:
If we know that N = 1, we should probably not allow calling the fused version on GPU. I'd prefer an error message over silently producing NaNs.

benbarsdell replied:
Unfortunately this can't easily be done in the Python layer because it only applies in training mode. Adding a check to build() causes a bunch of tests to fail (ones that don't use training mode). It would have to be an in-graph check, which I don't think there's much precedent for(?).

rmlarsen replied:
You can add a tf.assert in graph check here. This is not unprecedented and no different from adding a check in the kernel itself.

benbarsdell replied:
I've added an in-graph check using Assert.

benbarsdell replied:
It caused a bunch of other tests to fail. I'll have to look into them.

reedwm commented on tensorflow/python/keras/layers/normalization_test.py:
```diff
@@ -138,6 +138,70 @@ def test_batchnorm_convnet_channel_last(self):
+    np.testing.assert_allclose(np.std(out, axis=(0, 1, 2, 3)), 1.0, atol=1e-1)
+
+  @keras_parameterized.run_all_keras_modes
+  def test_batchnorm_convnet_channel_last_3d_fused_correctness(self):
```
I think there are no non-4D fused tests where the axis isn't the last element. Either change this test or the previous to have the axis be not the last element, or add a new test.

benbarsdell replied:
I've added channels-first versions of the two tests.

reedwm commented:
> I managed to root-cause the `layer_correctness_test` failure. There appears to be a bug [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L60) (also [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L229)) where the input is interpreted as `OutputT` instead of `InputT`.
> I'm not sure how to add a tensor typecast there, but just removing that piece of code does fix the problem.
> 
> Let me know how you want to deal with it.

Can you create a new PR to fix it? [This line](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L195) has an example of how to cast. If this is trickier than I'm imagining and you don't want to deal with this, I can also get someone internally to fix it.

benbarsdell commented:
> > I managed to root-cause the `layer_correctness_test` failure. There appears to be a bug [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L60) (also [here](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L229)) where the input is interpreted as `OutputT` instead of `InputT`.
> > I'm not sure how to add a tensor typecast there, but just removing that piece of code does fix the problem.
> > Let me know how you want to deal with it.
> 
> Can you create a new PR to fix it? [This line](https://github.com/tensorflow/tensorflow/blob/9c7ddffd970e8e72585c66c4a2c7cf8bb42ed608/tensorflow/core/kernels/redux_functor.h#L195) has an example of how to cast. If this is trickier than I'm imagining and you don't want to deal with this, I can also get someone internally to fix it.

Thanks, I tried that but unfortunately only Eigen tensors have a `cast` method; the TF Tensor type does not. At this stage it's probably easier for someone internal to fix it, unless there's some other very easy fix that I haven't identified.

benbarsdell committed:
More corrections for PR 40338

- Use keepdims=True to avoid needing to reshape.
- Remove unnecessary replacements of first dim with -1.
- Add channel-first versions of the two 3d tests.

reedwm reviewed (commented):
This change looks good, but I am worried about using Bessel's correction in more cases now. It's possible (but unlikely) we may break someone who currently relies on Bessel's correction not being applied.

I think the best solution is to change this layer to use Bessel's correction in the unfused case before submitting this change. Then we can extensively test the Bessel's correction change before submitting, potentially rolling it back if it breaks someone. Once submitted, this change is a lot safer since it doesn't affect correctness, only performance. I can work on using Bessel's correction in the unfused case since I can run all Google tests before submitting it.

Do you know why Bessel's correction for cuDNN is only used for the moving variance, not the internal variance the input is divided by? We copy this behavior in the CPU version of FusedBatchNorm as well.

gbaned commented:
@benbarsdell Can you please check @reedwm's comments and keep us posted. Thanks!

benbarsdell commented:
Regarding Bessel's correction at inference time but not training time, it seems that the original batchnorm paper indicated doing it this way; there is some discussion here (evidently it is the same way in PyTorch):
https://github.com/pytorch/pytorch/issues/1410
https://stats.stackexchange.com/questions/311074/batch-normalization-variance-calculation

It sounds like you are going to take care of the remaining items internally @reedwm. If I've misinterpreted that or you need me to do anything else just let me know.

reedwm commented:
Yes I will take care of using Bessel's correction for the nonfused implementation internally. Then we can merge this PR.

gbaned commented:
@reedwm Any update on this PR? Please. Thanks!

gbaned commented:
@reedwm Any update on this PR? Please. Thanks!

gbaned commented:
@benbarsdell  Can you please resolve conflicts? Thanks!

benbarsdell committed:
Merge upstream/master into fused-batchnorm-any-dims-and-axis

- Conflicts fixed in normalization.py and normalization_test.py in
  tensorflow/python/keras/layers/.

benbarsdell committed:
Check dims are valid in fused BatchNorm training

- This is an in-graph check required due to the use of Bessel's
  correction in fused batchnorm.
- Previously the layer would silently return invalid or NaN results
  if N <= 1.
- Also fixes a test that failed this condition (since it is now run with
  fused=True).

rmlarsen reviewed (changes requested):
Reed wanted to discuss a few design issues before approving.

gbaned commented:
@benbarsdell Can you please resolve conflicts? Thanks!

reedwm commented on tensorflow/python/keras/layers/normalization.py:
```diff
@@ -499,6 +524,37 @@ def _fused_batch_norm(self, inputs, training):
+    # The use of Bessel's correction in training mode imposes the requirement
+    # that the number of elements in the reduced dimensions is > 1.
+    check = control_flow_ops.Assert(
+          training == False or math_ops.reduce_prod(input_shape) > self.depth,
```
I think training can also be a tensor. The assert should be added to [this function](https://github.com/tensorflow/tensorflow/blob/807f17221d628b9fc722df5d99562f5ef126842c/tensorflow/python/keras/layers/normalization.py#L552) I think

reedwm replied:
Also probably better to also raise an error immediately if the shape and training boolean are statically known

benbarsdell commented:
It turns out that the reshapes added by this PR prevent the grappler layout optimizer from optimizing out all the NCHW<->NHWC transposes.

For this reason I think we will have to abandon this PR and use https://github.com/tensorflow/tensorflow/pull/42970 instead.
